### PR TITLE
Add pass thru mode

### PR DESCRIPTION
**Pull request by PeterC-DLS**
Refactor to pass fewer arguments, reduce attributes, encapsulate program specific slicer and modes, update tests, and update deprecated equals calls

### CODE REVIEW

File: ParProcCo/job_controller.py [hunk 1]
```diff
@@ -5,91 +5,83 @@
...
-    def run(self, data_slicer: SlicerInterface, number_jobs: int, processing_mode: SchedulerModeInterface,
-            aggregating_mode: SchedulerModeInterface, cluster_runner: Path,
-            jobscript_args: Optional[List] = None, aggregator_path: Optional[Path] = None, memory: str = "4G", cores: int = 6,
-            job_name: str = "ParProcCo") -> None:
+    def run(self, number_jobs: int, cluster_runner: Path,
+            jobscript_args: Optional[List] = None, aggregator_path: Optional[Path] = None,
+            memory: str = "4G", cores: int = 6, job_name: str = "ParProcCo") -> None:
 
-        slice_params = self.create_slices(data_slicer, number_jobs)
+        slice_params = self.program_wrapper.create_slices(number_jobs) if number_jobs > 1 else []
```
Comment (VictoriaBeilsten-Edmands) [inline ParProcCo/job_controller.py:31]: I do not think setting this to `[]` works:
`len(slice_params)` is used to set number of jobs in `SimpleProcessingMode` which is used in L94 of the `JobScheduler.run` method:
`job_indices = list(range(scheduler_mode.number_jobs))`. If `job_indices` is empty no job template are created and run.

Comment (VictoriaBeilsten-Edmands) [reply ParProcCo/job_controller.py:31]: Needs to use PassThruMode

Comment (PeterC-DLS) [reply ParProcCo/job_controller.py:31]: Yes, but `msm_cluster_submit` sets the wrapper to be `PassThruMode` when `jobs==1`

File: ParProcCo/passthru_processing_mode.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    def set_parameters(self, slice_params: List[slice]) -> None:
+        """Overrides SchedulerModeInterface.set_parameters"""
+        self.number_jobs = 1
+
+    def generate_output_paths(self, output_dir: Path, error_dir: Path, i: int) -> Tuple[str, str, str]:
+        """Overrides SchedulerModeInterface.generate_output_paths"""
+        std_out_fp = str(error_dir / f"std_out_{i}")
+        err_fp = str(error_dir / f"err_{i}")
+        return output_dir, std_out_fp, err_fp
```
Comment (VictoriaBeilsten-Edmands) [inline ParProcCo/passthru_processing_mode.py:20]: Returns output_dir as Path not str, so cannot create JobTemplate with args.

File: ParProcCo/job_controller.py (hunk 1 above)
Comment (VictoriaBeilsten-Edmands) [reply ParProcCo/job_controller.py:31]: in test_job_controller, `test_single_job_does_not_aggregate` doesn't pass.

Comment (PeterC-DLS) [reply ParProcCo/job_controller.py:31]: I thought I tested it!

Comment (PeterC-DLS) [reply ParProcCo/job_controller.py]: Missed commit my fix!

**Review by VictoriaBeilsten-Edmands — APPROVED** [review]

**Pull request merged**